### DISCUSSION TITLE
MindSearch只有除了web的搜索能力，可以做本地搜索吗？比如当做本地RAG，搭配搜索文档、向量等

### DISCUSSION THREAD

**Discussion by LIUKAI0815** [InternLM/MindSearch#41] [Announcements]

**Comment by Harold-lkk** [comment]
这部分我们没有实现，如果把多智能体中的 WebSearch 换成一个 RAG 系统应该是可以做到本地搜索的。 WebSearch 基本等价于利用网络知识做 RAG

**Reply by lzz999997** [reply]
请问我可以同时搜索数据库和网络吗？你们有微调模型吗？我想微调模型让他思考搜索数据库和网络

**Comment by datalee** [comment]
> 这部分我们没有实现，如果把多智能体中的 WebSearch 换成一个 RAG 系统应该是可以做到本地搜索的。 WebSearch 基本等价于利用网络知识做 RAG

其实不用了，直接只改search的地方就行

**Reply by qinchihongye** [reply]
老哥指明下方向，如果改本地搭建的搜索的话，应该改哪里

**Comment by LIUKAI0815** [comment]
@datalee 改成本地的知识库吗？

**Comment by datalee** [comment]
> @datalee 改成本地的知识库吗？

改成你本地知识库上构建的搜索api

**Reply by qinchihongye** [reply]
老哥，指明下大概改哪个地方

**Reply by datalee** [reply]
> 老哥，指明下大概改哪个地方

改这里：`lagent\actions\bing_browser.py`，模仿DuckDuckGoSearch类，新增个自己的xxx类；
then，在`mindsearch\agent\__init__.py`里36行：
`searcher_cfg=dict(
            llm=llm,
            # plugin_executor=ActionExecutor(
            #     BingBrowser(searcher_type='DuckDuckGoSearch',
            #                 topk=6,
            #                 api_key=os.environ.get('BING_API_KEY',
            #                                        'YOUR BING API')),`
的`searcher_type='DuckDuckGoSearch'`改成你的xxx

**Reply by qinchihongye** [reply]
我目前就这样改了，暂时能用了哈哈，感谢大佬指明方向，我直接将`lagent\actions\bing_browser.py`中的`DuckDuckGoSearch`
类改了，我增加了下面的代码
```python
def modify_ddsg_response(query:str,debug=False):
    """自己的搜索接口"""
    url = 'http://******' 
    body = {
        "query":query,
        "datasetIds":["661f9c02e50601c600e26eb9"],
        "topN": 9,
        "simThreshold": 0.0
    }
    res = requests.post(url,json=body).json()
    conv_links,content_list = res['conv_links'][0:9],res['content_list'][0:9] # 最多只选 9 个
    result = []
    for d in content_list:
        r_d = {}
        r_d['title'] = d['topic']
        r_d['href'] = d['url']
        r_d['body'] = d['content']
        result.append(r_d)
    # result 的返回形式为 [{'title':标题,'href':url地址,'body':摘要或者内容都行},{},...]
    return result
```
修改了`_call_ddgs`
主要就是吧response凑成原有的返回形式
```python
response = modify_ddsg_response(query.strip("'"))
```
暂时反正是能用了，后面过几天有时间再再试试直接把BingBrowser类改写一下
感谢大佬指明方向

**Reply by wangtianqi1993** [reply]
hi 你好，请教下你目前返回的结果中 r_d['href'] = d['url'] ， 是真正的url吗，后面的BingBrowser类中的select方法能直接对这个URL提取全部内容不，否则不是真正url并且不修改select中的代码是不是会报错

**Reply by datalee** [reply]
> hi 你好，请教下你目前返回的结果中 r_d['href'] = d['url'] ， 是真正的url吗，后面的BingBrowser类中的select方法能直接对这个URL提取全部内容不，否则不是真正url并且不修改select中的代码是不是会报错

是真的，哈哈哈哈，select是根据summ字段进行筛选，
